README.md

Обсуждение и скриншоты тут - http://forum.imagecms.net/viewtopic.php?id=4988

### Конструктор форм - xforms для ImageCMS (на Codeigniter, старого образца)

## Обратите внимание!
Файлы модуля должны находиться в папке /application/modules/xforms

Модуль корректно работает с **ImageCMS** версий **от 4.9 до 4.12.1**

## Что сделано?
* Отправка почты на несколько email адресов.
* Отправка почты частично работает через модуль "Управление email-уведомлениями"
* Все формы на AJAX и отправляются без перезагрузки страницы.
* Поля формы можно включать/выключать.
* Нормально работающий дизайн админ панели для формы.
* Автоопределение города посетителя через сервис 
* Функция прямого доступа к форме через УРЛ. (не путать с запретом доступа к модулю через УРЛ)
* Функция прикрепления файлов к форме. (нужно создать поле "загрузка файла")
* Форму можно встроить как виджет.
* Функция "защитный код (каптча)". Если в форме есть ошибки, капча автоматически обновляется.
* Уведомление об ошибках "на лету", т.е. если вы ввели что то неверно, будет уведомление об ошибке (после нажатия на кнопку отправить)
* Есть такой мифический тип поля, "разделитель". Нужен например для визуального отделения мелких групп полей. Не стал заморачиваться с fieldset и legend.
* Вырезан лишний JS нотификации при отправке письма, и добавлены пользовательские функции. Подробнее тут - http://forum.imagecms.net/viewtopic.php?pid=26212#p26212 (все по разному хотят выводить нотификации и другие действия с формой после её отправки)

## Баги
* При загрузке файла PSD, всегда пишет - "не верный тип файла", причем пробовал заменить mimes.php из codegniter 3 версии, все равно ругается. Пока не думал как решать.
* Если на странице больше двух форм, и включена каптча в формах, то в первой форме будет не верное изображение каптчи

### Что в планах? (в порядке убывания приоритетов)
* JS валидация в соответствии с выставленными условиями проверки `(valid_email|max_length[255]|min_length[1]|numeric итд)` - http://rickharrison.github.io/validate.js/
* Добавить дефолтные стили. (возможно сделать опцию использования дефолтных стилей)
* Добавление поля Time, data и time data. (php валидация уже сделана) (На основе Js - datetimepicker)
* Добавить поддержку reCaptcha
* Мультиязычность

## Обновления:
см. тут https://github.com/pulse-studio/xforms/blob/master/changelog.md

changelog.md
##### UPD: 04/05/19 - Версия 3.0.9
* (add) Определение города посетителя по IP, через сервис  `ip-api.com`
* (add) Вывод IP и информации о городе в письме.


##### UPD: 04/05/19 - Версия 3.0.8
* (add) тип поля `hidden` (input type="hidden")

##### UPD: 07/08/18 - Версия 3.0.7
* добавлен тип поля "tel" (телефон)
* добавлен тип поля "email"
* небольшие правки в *.tpl вывода формы (обновите в своем проекте, если будите обновляться)

##### UPD: 07/08/18 - Версия 3.0.6
* Форматирование php кода.
* добавлен функционал вывода кастомных ошибок для полей.

##### UPD: 07/08/18 - Версия 3.0.5.1
* (bugfix) Ошибка создания формы из-за наличия отсутствующих полей в БД subject и email.

##### UPD: 09/06/18 - Версия 3.0.4
* Обновлено руководство, добавлена страница со списком доступных фукнций валидации полей - https://github.com/pulse-studio/xforms/wiki/%D0%A1%D0%BF%D0%B8%D1%81%D0%BE%D0%BA-%D0%B4%D0%BE%D1%81%D1%82%D1%83%D0%BF%D0%BD%D1%8B%D1%85-%D1%84%D1%83%D0%BD%D0%BA%D1%86%D0%B8%D0%B9-%D0%B2%D0%B0%D0%BB%D0%B8%D0%B4%D0%B0%D1%86%D0%B8%D0%B8-%D0%BF%D0%BE%D0%BB%D0%B5%D0%B9

##### UPD: 28/04/18 - Версия 3.0.4
* Иправление бага с checkbox, передавались неверные значения

##### UPD: 04/04/18 - Версия 3.0.3
* Доработан перевод админ панели и клиентской части
* Удален лишний код в admin.php
* Исправлена ошибка xss_clean у описания и сообщения успешной отправки, при редактировании формы.

##### UPD: 04/04/18 - Версия 3.0.2
* Удалено лишнее поле subject в БД
* Частичный перевод админ панели
* Добавлена ссылка на редактирования cmsemail в списке форм, в админ панели
* Добавлена обертка для показа формы по прямому УРЛ. Нужно для создания нужных HTML тегов, не нарушая стилистику сайта. Менять `\application\modules\xforms\templates\wrapper.tpl`

##### UPD: 03/04/18 - Версия 3.0.1
* (bugfix) исправлен баг с maxlength у textarea

##### UPD: 22/03/18 - Версия 3.0 
* Частичная интеграция со стандартным модулем cmsemail, для отправки писем.
* (bugfix) При отправке писем с вложением на несколько адресов администратора, дублировались вложения. На 2 емайл 2 раза, на 3, 3 итд.

##### UPD: 22/03/18 - Версия 2.9
* Вырезан функционал сохранения сообщений в админ панели (на мой взгляд он лишний)
* Улучшен код по заморозке/разморозке кнопки отправки формы при загрузке файлов
* Добавлены новые варианты валидации, Телефон - callback_valid_phone, Дата - callback_valid_date, Время - callback_valid_time.
* (bugfix) в поле "файл". Если на странице было 2 или более форм с файловыми полями, файлы загружались несколько раз
* (bugfix) При удалении формы теперь удаляются шаблоны cmsemail для данной формы

##### UPD: 06/06/17
* Сжаты JS для загрузки файлов. (Для более быстрой загрузки сайта)

##### UPD: 30/03/17 - Версия 2.4.2
* Мелкие правки в tpl
* Созданы функции валидации для полей типа (телефон, дата, время). В дальшейшем будут добавлены новые типы полей.

##### UPD: 21/02/17 - Версия 2.4.0
* Для обновления БД зайдите в админпанель и перейдите по адресу - `/admin/components/cp/xforms/update_2_4/`
* Внесены правки в шаблон отображения формы на клиентской части
* Теперь отправка почты частично работает через модуль "Управление email-уведомлениями". Частично, потому что - https://github.com/imagecms/ImageCMS/issues/103
* Небольшие правки в tpl админ-панели
* Теперь есть возможность выбрать что делать с загруженными файлами. Доступно 3 варианта. 1)Сохранить на сервере. 2)Отправить вложением и удалить с сервера. 3) Отправить вложением и сохранить на сервере. Кому-то будет полезно.
* Теперь сообщения добавляются в админ панель.

##### UPD: 30/01/17
* Исправлена ошибка загрузки скриптов для поля "загрузка файла"

##### UPD: 28/11/16 - Версия 2.3.1
* Загрузка файлов вынесена в отдельный файл xforms_files.js
* FIX: Если в форме нет полей с файлами, то не подгружает скрипты для загрузки файлов.
* Исправления в шаблоне админки.
* Вырезан лишний JS нотификации при отправке письма, добавлен вызов пользовательских функции. Пример тут - http://forum.imagecms.net/viewtopic.php?pid=26212#p26212 (все по разному хотят выводить нотификации и другие действия с формой после её отправки).

##### UPD: 10/11/16
* Ошибки в xforms.js

##### UPD: 04/11/16
* Исправлена ошибка в tpl письма + добавлена функция скачивания файлов

##### UPD: 19/10/16 - Версия 2.3
* Добавил новый тип поля, для формы, - "загрузка файла". Можно загружать файлы через ajax и выбирать какие типы файлов разрешено загружать.
* Теперь можно сделать первое значение `select/checked/radio` выбранным по умолчанию.
* Косметические исправления при редактировании поля в админке.
* checkbox работает как надо

Для обновления модуля с версии 2.0 до 2.3 замените все файлы модуля, а так-же зайдите в админ панель и перейдите по адресу - `/admin/components/cp/xforms/update_2_3/`

##### UPD: 11/06/16
* Теперь работают типы поля checkbox и исправил ошибки в radio
* Модуль будет по умолчанию отправлять письсма через SMTP если он у вас настроен.

##### UPD: 11/05/16
* При создании поля, его позиция по умолчанию = ID поля, если она не указана. (поле помещается в конец списка)

admin.php
<?php

use CMSFactory\assetManager;
use cmsemail\email;
use Components;


if (!defined('BASEPATH')) {
    exit('No direct script access allowed');
}


/**
 * @property Xforms_model xforms_model
 */
class Admin extends BaseAdminController
{

    public function __construct()
    {
        parent::__construct();
        $lang = new MY_Lang();
        $lang->load('xforms');
        $this->load->model('xforms_model');
        $this->load->library('form_validation');
    }

    /**
     * Изменение статуса, видим или не видим, для поля формы
     * @param int $field_id
     */
    public function change_visible($field_id)
    {
        $field = $this->xforms_model->get_field($field_id);

        if ($field['visible'] == 1) {
            $field['visible'] = 0;
        } elseif ($field['visible'] == 0) {
            $field['visible'] = 1;
        }

        $this->xforms_model->update_field($field['id'], $field);
        showMessage(lang('Status change success', 'xforms'));
    }

    /**
     * @param string $data
     * @return bool
     */
    public function check_emails($data)
    {
        return $this->form_validation->valid_emails($data);
    }

    /**
     * Удаление полей формы
     */
    public function delete_fields()
    {
        $this->xforms_model->delete_fields($this->input->post('id'));
    }

    /**
     * Удаление формы
     */
    public function delete_form()
    {
        if (count($this->input->post()) > 0) {
            $this->db->where('id', (int)$this->input->post('id'))->delete('xforms');
            $this->db->where('fid', (int)$this->input->post('id'))->delete('xforms_field');

            // Удаляем email шаблон
            $email_template = $this->db->select('id')->where('name', 'xforms_send_form_' . $this->input->post('id'))->get('mod_email_paterns')->row_array();
            $this->db->where('id', (int)$email_template['id'])->delete('mod_email_paterns');
            $this->db->where('id', (int)$email_template['id'])->delete('mod_email_paterns_i18n');
        }
    }

    /**
     * Работа с полями формы
     * @param null|int $fid
     * @param null|string $field
     */
    public function field($fid = null, $field = null)
    {
        if ($this->input->post('type')) {
            $this->form_validation->set_rules('value', 'Значение', 'trim|xss_clean');
            $this->form_validation->set_rules('desc', 'Описание', 'trim|xss_clean|max_length[255]');
            $this->form_validation->set_rules('validation', 'Валидация', 'trim|xss_clean');
            $this->form_validation->set_rules('operation', 'Операции', 'trim');
            $this->form_validation->set_rules('position', 'Позиция', 'trim|xss_clean|numeric');
            $this->form_validation->set_rules('maxlength', 'Максимум символов', 'trim|xss_clean|numeric');
            $this->form_validation->set_rules('allowed_types', 'Типы файлов', 'trim|xss_clean|max_length[500]');

            if ($this->form_validation->run($this) == FALSE) {
                showMessage(validation_errors(), false, 'r');
            } else {
                $data = [
                    'fid' => $fid,
                    'type' => $this->input->post('type'),
                    'label' => $this->input->post('name'),
                    'value' => $this->input->post('value'),
                    'error_message' => $this->input->post('error_message'),
                    'desc' => $this->input->post('desc'),
                    'operation' => $this->input->post('operation'),
                    'position' => $this->input->post('position'),
                    'maxlength' => $this->input->post('maxlength'),
                    'checked' => $this->input->post('check'),
                    'disabled' => $this->input->post('disable'),
                    'require' => $this->input->post('required'),
                    'validation' => $this->input->post('validation'),
                    'visible' => $this->input->post('visible'),
                    'allowed_types' => $this->input->post('allowed_types'),
                ];

                if (!$field) {
                    $field_id = $this->xforms_model->add_field($data);

                    if (!$data['position']) {
                        $data['position'] = $field_id;
                        $this->xforms_model->update_field($field_id, $data);
                    }

                    showMessage(lang('Field created', 'xforms'));
                    $path = '/admin/components/cp/xforms/field/' . $fid . '/' . $field_id;
                } else {

                    $this->xforms_model->update_field((int)$field, $data);
                    showMessage(lang('Changes has been saved', 'xforms'));
                    $path = '/admin/components/cp/xforms/field/' . $fid . '/' . $field;
                }

                if ($this->input->post('action') == 'exit') {
                    $path = '/admin/components/cp/xforms/fields/' . $fid;
                }

                pjax($path);
            }
        } else {
            if ($field) {
                assetManager::create()
                    ->setData('field', $this->xforms_model->get_field((int)$field))
                    ->setData('fid', $fid)
                    ->renderAdmin('field');
            } else {
                assetManager::create()->setData('fid', $fid)->renderAdmin('field');
            }
        }
    }

    /**
     * EDIT and ADD field for form
     * @param int $id
     */
    public function fields($id)
    {
        assetManager::create()
            ->setData('fields', $this->xforms_model->get_form_fields($id))
            ->setData('form_name', $this->xforms_model->get_form_name($id))
            ->setData('form_id', $id)
            ->renderAdmin('fields');
    }

    /**
     * CREATE and EDIT form
     * создание / редактирование формы
     * @param null|int $id
     */
    public function form($id = null)
    {
        // Сохраняем
        if ($this->input->post()) {

            $this->form_validation->set_rules('page_title', lang('Title form', 'xforms'), 'trim|required|min_length[1]|max_length[255]');
            $this->form_validation->set_rules('page_url', lang('URL form', 'xforms'), 'alpha_dash|least_one_symbol|max_length[255]');
            $this->form_validation->set_rules('desc', lang('Description', 'xforms'), 'trim');
            $this->form_validation->set_rules('good', lang('Successful message', 'xforms'), 'trim|required');

            // Валидируем заголовок темы и email только когда форма еще не создана
            if (!$id) {
                $this->form_validation->set_rules('subject', lang('Email subject', 'xforms'), 'trim|xss_clean|required|min_length[1]|max_length[255]');
                $this->form_validation->set_rules('email', 'Email', 'callback_check_emails');
            }

            if ($this->form_validation->run($this) == FALSE) {
                showMessage(validation_errors(), false, 'r');
            } else {
                // Собираем данные
                $data = [
                    'title' => $this->input->post('page_title'),
                    'url' => $this->input->post('page_url'),
                    'desc' => $this->input->post('desc'),
                    'success' => $this->input->post('good'),
                    'captcha' => $this->input->post('captcha'),
                    'direct_url' => $this->input->post('direct_url'),
                    'user_message_active' => $this->input->post('user_message_active'),
                    'action_files' => $this->input->post('action_files'),
                ];

                $email_data = [
                    'subject' => $this->input->post('subject'),
                    'email' => $this->input->post('email'),
                ];

                // Создаем / сохраняем
                if (isset($id) AND $this->xforms_model->update_form($id, $data)) {

                    showMessage(lang('Changes has been saved', 'xforms'));
                    $path = '/admin/components/cp/xforms/form/' . $id;
                } else {
                    $id = $this->xforms_model->add_form($data);
                    showMessage('Готово', 'Форма добавлена');
                    $path = '/admin/components/cp/xforms/form/' . $id;

                    // Добавим шаблон "xforms_send" для отправки почты через модуль cmsemail
                    $this->load->dbforge();
                    $email_paterns = [
                        'name' => 'xforms_send_form_' . $id,
                        'from' => 'Администрация сайта',
                        'from_email' => '',
                        'type' => 'HTML',
                        'patern' => '',
                        'user_message_active' => 0,
                        'admin_message_active' => 1,
                        'admin_email' => $email_data['email']
                    ];
                    $this->db->insert('mod_email_paterns', $email_paterns);
                    $email_patterns_id = $this->db->insert_id();

                    $email_paterns_i18n = [
                        'id' => $email_patterns_id,
                        'locale' => 'ru',
                        'theme' => $email_data['subject'] ? $email_data['subject'] : 'ImageCMS - Отправка формы - ' . $id,
                        'user_message' => 'Здравствуйте. Мы свяжемся с вами в ближайшее время.',
                        'admin_message' => '<p>Пришла заявка.</p><p>$message$</p>',
                        'description' => 'Отправка формы xforms - ' . $data['title'],
                        'variables' => 'a:1:{s:9:"$message$";s:46:"Список заполненных полей";}'
                    ];
                    $this->db->insert('mod_email_paterns_i18n', $email_paterns_i18n);

                }

                if ($this->input->post('action') == 'close') {
                    $path = '/admin/components/cp/xforms';
                }

                pjax($path);

            }
        } else {
            // Show form
            if (isset($id)) {
                assetManager::create()
                    ->setData('form', $this->xforms_model->get_form($id))
                    ->setData('cmsemail', email::getInstance()->cmsemail_model->getPaternSettings('xforms_send_form_' . $id))
                    ->setData('fields', $this->xforms_model->get_form_fields($id, ['visible' => 1, 'type' => 'text']))
                    ->renderAdmin('form');
            } else {
                assetManager::create()->renderAdmin('form');
            }
        }
    }

    /**
     * Show list forms in admin panel
     */
    public function index()
    {
        $xforms = $this->xforms_model->get_forms();
        foreach ($xforms as $key => $f) {
            $xforms[$key]['cmsemail_id'] = $this->db->select('id')->where('name', 'xforms_send_form_' . $f['id'])->get('mod_email_paterns')->row_array()['id'];
        }

        assetManager::create()->setData('forms', $xforms)->renderAdmin('forms');
    }

    /**
     * Sort fields in the form
     */
    public function update_positions()
    {
        $positions = $this->input->post('positions');

        foreach ($positions as $key => $value) {
            $this->db->where('id', (int)$value)->set('position', $key)->update('xforms_field');
        }

        showMessage('Позиция обновлена');
    }


    /**
     * Обновление модуля
     */
    public function update()
    {
        /**
         * с 3.0 версии до 3.0.6
         * Добавим колонку с сообщением об ошибке
         */
        $xforms_field = [
            'error_message' => [
                'type' => 'varchar',
                'constraint' => 255,
            ]
        ];
        $this->load->dbforge();
        $this->dbforge->add_column('xforms_field', $xforms_field);
    }
}